bidict/_iter.py

```diff
@@ -9,13 +9,15 @@
 
 from collections.abc import Mapping
 from operator import itemgetter
+import typing as t
 
 from ._typing import KT, VT, IterItems, MapOrIterItems
 
 
 def iteritems_mapping_or_iterable(arg: MapOrIterItems[KT, VT]) -> IterItems[KT, VT]:
     """Yield the items in *arg* based on whether it's a mapping."""
-    yield from arg.items() if isinstance(arg, Mapping) else arg
+    it = iter(arg.items() if isinstance(arg, Mapping) else arg)
+    return t.cast(IterItems[KT, VT], it)
```

Comment:
I'd have to look at this more closely, but it's possible there is a bug where pyright substitutes the type variables in the `MapOrIterItems` alias incorrectly. To investigate further I'd add `reveal_type()` calls in a few branches.

Reply:
Thanks @JelleZijlstra, does this help?
```diff
 def iteritems_mapping_or_iterable(arg: MapOrIterItems[KT, VT]) -> IterItems[KT, VT]:
     """Yield the items in *arg* based on whether it's a mapping."""
-    it = iter(arg.items() if isinstance(arg, Mapping) else arg)
-    return t.cast(IterItems[KT, VT], it)
+    if isinstance(arg, t.Mapping):
+        foo = arg.items()
+        reveal_type(foo)
+    else:
+        foo = arg
+        reveal_type(foo)
+    yield from foo
```
```console
❯ pyright bidict/_iter.py
Loading configuration file at /Users/jab/src/bidict/pyrightconfig.json
Assuming Python version 3.10
Assuming Python platform Darwin
stubPath /Users/jab/src/bidict/typings is not a valid directory.
Searching for source files
Found 1 source file
/Users/jab/src/bidict/bidict/_iter.py
  /Users/jab/src/bidict/bidict/_iter.py:20:21 - information: Type of "foo" is "ItemsView[KT@iteritems_mapping_or_iterable, VT@iteritems_mapping_or_iterable] | ItemsView[Tuple[KT@iteritems_mapping_or_iterable, VT@iteritems_mapping_or_iterable], Unknown]"
  /Users/jab/src/bidict/bidict/_iter.py:23:21 - information: Type of "foo" is "Iterable[Tuple[KT@iteritems_mapping_or_iterable, VT@iteritems_mapping_or_iterable]]"
  /Users/jab/src/bidict/bidict/_iter.py:24:16 - error: Expression of type "tuple[KT@iteritems_mapping_or_iterable, VT@iteritems_mapping_or_iterable] | tuple[Tuple[KT@iteritems_mapping_or_iterable, VT@iteritems_mapping_or_iterable], Unknown]" cannot be assigned to yield type "Tuple[KT@iteritems_mapping_or_iterable, VT@iteritems_mapping_or_iterable]"
    Type "tuple[KT@iteritems_mapping_or_iterable, VT@iteritems_mapping_or_iterable] | tuple[Tuple[KT@iteritems_mapping_or_iterable, VT@iteritems_mapping_or_iterable], Unknown]" cannot be assigned to type "Tuple[KT@iteritems_mapping_or_iterable, VT@iteritems_mapping_or_iterable]"
      Tuple entry 1 is incorrect type
        Type "Tuple[KT@iteritems_mapping_or_iterable, VT@iteritems_mapping_or_iterable]" cannot be assigned to type "KT@iteritems_mapping_or_iterable" (reportGeneralTypeIssues)
1 error, 0 warnings, 2 informations
Completed in 0.664sec
```

Reply:
Thanks! I think line 20 tells us what's going on: pyright thinks that a `MapOrIterItems[KT, VT]` that is a `Mapping` can also be a `Mapping[tuple[KT, VT], Any]`, because that last type is compatible with `IterItems[KT, VT] == Iterable[tuple[KT, VT]]`. That's technically correct, but not very helpful here. I can't immediately think of a good workaround.

Reply:
There are two issues here. First, if you use a `yield`, the function is a generator, and its return type should reflect that. That means the proper return type would be `t.Generator[IterItems[KT, VT], None, None]`.

The second issue is the same one as above with the `isinstance` narrowing of `MapOrIterItems`. You could solve that the same way as I suggested above.

Reply:
> the function is a generator, and its return type should reflect that

I'm not following this. The function happens to be implemented via a generator currently, but promising that in the return type hint would needlessly tie the interface to that implementation. It also gives callers license to use the other generator APIs (`send`, `throw`, `close`), which doesn't make sense here. Better to only promise an Iterable is returned, which may be implemented via a generator, but that's just an implementation detail that callers shouldn't count on. Does that make sense?

> The second issue is the same one as above with the isinstance narrowing of MapOrIterItems. You could solve that the same way as I suggested above.

Thanks for explaining. Unfortunately that solution wasn't preferable for the case above, and I think it's not a good fit here for similar reasons.

Reply:
Ignore what I said about the return type needing to be a `Generator`. An `Iterable` or `Iterator` is fine too.

Reply:
Thanks for confirming about `Generator`.

I'm going to leave the original implementation here (`yield from arg.items() if isinstance(arg, Mapping) else arg`), and just add a `pyright: ignore`, since this seems like a case where the code is both correct and most maintainable as written, at least given the typing functionality available in Python 3.7.

I'll leave this thread unresolved in case @JelleZijlstra wants to take another look at why mypy is not flagging this line.

Reply:
Seems like the same issue as https://github.com/jab/bidict/pull/242#discussion_r825380476